dask_cuda/cuda_worker.py

```diff
@@ -0,0 +1,253 @@
+                                   "is leaking memory?  Process memory: %s -- "
+                                   "Worker memory limit: %s",
+                                   format_bytes(proc.memory_info().rss),
+                                   format_bytes(self.memory_limit))
+                    break
+                k, v, weight = self.data.host.fast.evict()
+                del k, v
+                total += weight
+                count += 1
+                yield gen.moment
+                memory = proc.memory_info().rss
+                if total > need and memory > target:
+                    # Issue a GC to ensure that the evicted data is actually
+                    # freed from memory and taken into account by the monitor
+                    # before trying to evict even more data.
+                    self._throttled_gc.collect()
+                    memory = proc.memory_info().rss
+            if count:
+                logger.debug("Moved %d pieces of data and %s bytes to disk",
+                             count, format_bytes(total))
+
+        self._memory_monitoring = False
+        raise gen.Return(total)
+
+    @gen.coroutine
+    def device_memory_monitor(self):
+        """ Track this process's memory usage and act accordingly
+
+        If we rise above (device_memory_spill_fraction * memory_limit) of
+        device memory use, start dumping data to disk. The default value
+        for device_memory_spill_fraction is 0.7, defined via configuration
+        'cuda.worker.device-memory.target'.
+
+        If we rise above (device_memory_pause_fraction * memory_limit) of
+        device memory use, stop execution of new tasks. The default value
+        for device_memory_pause_fraction is 0.8, defined via configuration
+        'cuda.worker.device-memory.pause'.
+        """
+        if self._memory_monitoring:
+            return
+        self._device_memory_monitoring = True
+        total = 0
+        memory = get_device_used_memory()
+        frac = memory / self.device_memory_limit
+
+        # Pause worker threads if device memory use above
+        # (self.device_memory_pause_fraction * 100)%
+        self._paused = self._check_for_pause(frac, self.device_memory_pause_fraction,
+                                             memory, self.device_memory_limit,
+                                             self._device_paused, None,
+                                             "Worker's CUDA device")
+        self.paused = (self._paused or self._device_paused)
+
+        # Dump device data to host if device memory use above
+        # (self.device_memory_spill_fraction * 100)%
+        if (self.device_memory_spill_fraction
+                and frac > self.device_memory_spill_fraction):
+            target = self.device_memory_limit * self.device_memory_target_fraction
+            count = 0
+            while memory > target:
+                if not self.data.device.fast:
```

Comment:
I have a question regarding , what is the data that is stored in `` CUDAWorker.data.device.fast:``? 

I was trying to persist a  data frame  bigger than the gpu memory but individual chunks comfortably fit in device memory but it seems to get paused and throw the following warning on both workers and get paused:

```python
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host. Perhaps some other process is leaking memory? Process memory: 13.76 GB -- Worker memory limit: 17.07 GB
```

Also, is there a way to access the ``CUDAWorker`` object  from the ``client``/``cluster`` object. I tried the ``CUDAWorker`` at ```client.cluster.workers[0].Worker``` but it does not seem contain the ```self.data```  attribute and other attributes.

Reply:
> I have a question regarding , what is the data that is stored in ` CUDAWorker.data.device.fast:`?

`CUDAWorker.data.device.fast` is a dict that stores `dask.array` device chunks. Similarly `CUDAWorker.data.device.slow` (`== CUDAWorker.data.host.fast`) is a dict that stores `dask.array` host chunks or device chunks that have been spilled to host, and `CUDAWorker.data.host.slow` contains the path to all chunks on disk.

> I was trying to persist a data frame bigger than the gpu memory with individual chunks that comfortably fit in device memory but it seems to get paused and throw the following warning on both workers and get paused:

IMO, this is one of the trickiest parts. We can't guarantee that `CUDAWorker` has full control over device memory, depending on what you're using to create/load your data, it may have its own memory pool, for example. This is why I had to explicitly disable CuPy's memory pooling to write a reliable test in https://github.com/rapidsai/dask-cuda/pull/35/files#diff-fa83f19df929827ec523fd00a21599b8R38.

> Also, is there a way to access the `CUDAWorker` object from the `client`/`cluster` object. I tried the `CUDAWorker` at `client.cluster.workers[0].Worker` but it does not seem contain the `self.data` attribute and other attributes.

I think this isn't possible, why would you like to access that directly? I don't know if this is by design or is just something that was never implemented, @mrocklin could you clarify this?

Reply:
>> Also, is there a way to access the CUDAWorker object from the client/cluster object. 

> I think this isn't possible

Correct.  The workers are in separate processes, so there is no way to access them from Python.  You can ask Dask to run functions on them to inspect state if you like with 

```python
def f(dask_worker):
    return len(dask_worker.data.device.fast)

client.run(f)
```

(See the run docstring for more information).

You could also try the `start_ipython_workers` function with the keyword `qtconsole=True`, but I woudln't be surprised if it has fallen out of use with recent Tornado library upgrades.

Reply:
Note that when @pentschev says  'dask.array device chunks" he also means any piece of GPU allocated data, which could be a dask array device chunk as he's dealing with, or a cudf dataframe as you're dealing with.

Reply:
Thanks a lot @mrocklin  for the function to inspect state. I used your function to debug whats happening. It seems that  the data is being evicted from  `data.device.fast` but the gpu memory is still not being cleared.  

```python 
('tcp://172.17.0.2:44914', {'data.device.fast': 14, 'data.device.slow': 0})
('tcp://172.17.0.2:44914', {'data.device.fast': 15, 'data.device.slow': 0})
('tcp://172.17.0.2:44914', {'data.device.fast': 3, 'data.device.slow': 14})
('tcp://172.17.0.2:44914', {'data.device.fast': 3, 'data.device.slow': 14})
('tcp://172.17.0.2:44914', {'data.device.fast': 0, 'data.device.slow': 18})
('tcp://172.17.0.2:44914', {'data.device.fast': 0, 'data.device.slow': 18})
```
Error:
```
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 12.08 GB -- Worker memory limit: 17.07 GB
```

May be the del here is not clearing memory. Don't know.  https://github.com/rapidsai/dask-cuda/pull/35/files#diff-c87f0866b277f959dc7c5d1e4b0ff015R243

Will add a small minimal reproducible example here soon.

Reply:
Probably worth asking cudf devs at this point.  @kkraus14 is there any reason why, after a bunch of cudf dataframe had been deleted, we would still detect a bunch of GPU memory being used?

Reply:
I suppose this is coming from [RMM](https://github.com/rapidsai/rmm). You could try [disabling the memory pool](https://github.com/rapidsai/rmm#handling-rmm-options-in-python-code) to see if that helps.

Reply:
Minimal Working Example of the pause i am facing. 

Code :
```python
# import required libs
import time
from dask.distributed import Client, wait
from dask_cuda import LocalCUDACluster
import dask_cudf

# Use dask-cuda to start one worker per GPU on a single-node system
cluster = LocalCUDACluster(ip='0.0.0.0',n_workers=1,device_memory_limit='5000 MB')
client = Client(cluster)
# print client info
print(client)


# Code to simulate_data

def generate_file(output_file,rows=100):
    with open(output_file, 'wb') as f:
        f.write(b'A,B,C,D,E,F,G,H,I,J,K\n')
        f.write(b'22,697,56,0.0,0.0,0.0,0.0,0.0,0.0,0,0\n'*rows)
        f.close()

# generate the test file 
output_file='test.csv'
generate_file(output_file,rows=25_000_000)


# reading it using dask_cudf
df = dask_cudf.read_csv(output_file,chunksize='100 MiB')


# Persisting it to disk
df = df.persist()

## Inspect State of the dask worker
def get_data(dask_worker):
    return {'data.device.fast':len(dask_worker.data.device.fast),
           'data.device.slow':len(dask_worker.data.device.slow),
           }
while True:
    time.sleep(1)
    for d in client.run(get_data).items():
        print(d)
```
Warning it gives out :
```python
distributed.worker - WARNING - Worker's CUDA device is at 82% memory usage. Pausing worker.  Process memory: 4.11 GB -- Worker memory limit: 5.00 GB

```


Client Trace:
```python
<Client: scheduler='tcp://172.17.0.2:44892' processes=1 cores=1>
('tcp://172.17.0.2:38393', {'data.device.fast': 0, 'data.device.slow': 0})
('tcp://172.17.0.2:38393', {'data.device.fast': 0, 'data.device.slow': 0})
('tcp://172.17.0.2:38393', {'data.device.fast': 1, 'data.device.slow': 0})
('tcp://172.17.0.2:38393', {'data.device.fast': 2, 'data.device.slow': 0})
('tcp://172.17.0.2:38393', {'data.device.fast': 4, 'data.device.slow': 0})
('tcp://172.17.0.2:38393', {'data.device.fast': 5, 'data.device.slow': 0})
('tcp://172.17.0.2:38393', {'data.device.fast': 6, 'data.device.slow': 0})
('tcp://172.17.0.2:38393', {'data.device.fast': 1, 'data.device.slow': 7})
('tcp://172.17.0.2:38393', {'data.device.fast': 0, 'data.device.slow': 9})
('tcp://172.17.0.2:38393', {'data.device.fast': 0, 'data.device.slow': 9})
('tcp://172.17.0.2:38393', {'data.device.fast': 0, 'data.device.slow': 9})
('tcp://172.17.0.2:38393', {'data.device.fast': 0, 'data.device.slow': 9}) ....
```

Reply:
> I suppose this is coming from [RMM](https://github.com/rapidsai/rmm). You could try [disabling the memory pool](https://github.com/rapidsai/rmm#handling-rmm-options-in-python-code) to see if that helps.

Thanks for this. Have not looked at rmm before but will try it out and let you know if this goes away.

Reply:
@VibhuJawa I found a couple of bugs that should be fixed now. The first one was, despite printing that the worker was pausing due to high CUDA memory usage, it didn't. The second was on printing the resume message, which would always print that the worker resumed due to the host memory being under the pause limit.

Nevertheless, your example was working by some luck, otherwise you would keep on seeing warning messages at regular intervals, not only the output you see on the client trace without warnings in between printing `get_data`.

Using the latest commit, this is what I see when `device_memory_limit='5000 MB'`:

```
<Client: scheduler='tcp://10.21.85.118:32995' processes=1 cores=1>
distributed.worker - WARNING - Worker's CUDA device is at 82% memory usage. Pausing worker.  Process memory: 4.14 GB -- Worker memory limit: 5.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 4.14 GB -- Worker memory limit: 5.00 GB
distributed.worker - WARNING - Worker is at 1% memory usage. Resuming worker. Process memory: 171.00 MB -- Worker memory limit: 16.48 GB
('tcp://10.21.85.118:33481', {'data.device.fast': 2, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 6, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
('tcp://10.21.85.118:33481', {'data.device.fast': 10, 'data.device.slow': 0})
```

In the end, the client/worker are just idle, with nvidia-smi reporting usage of 2953 MB.

If now I set `device_memory_limit='4000 MB'`, I have a different situation:

```
<Client: scheduler='tcp://10.21.85.118:44071' processes=1 cores=1>
distributed.worker - WARNING - Worker's CUDA device is at 103% memory usage. Pausing worker.  Process memory: 4.14 GB -- Worker memory limit: 4.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 4.14 GB -- Worker memory limit: 4.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 74% memory usage. Resuming worker. Process memory: 2.97 GB -- Worker memory limit: 4.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 2.97 GB -- Worker memory limit: 4.00 GB
('tcp://10.21.85.118:43265', {'data.device.fast': 1, 'data.device.slow': 0})
('tcp://10.21.85.118:43265', {'data.device.fast': 6, 'data.device.slow': 0})
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.16 GB -- Worker memory limit: 4.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.10 GB -- Worker memory limit: 4.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.10 GB -- Worker memory limit: 4.00 GB
('tcp://10.21.85.118:43265', {'data.device.fast': 0, 'data.device.slow': 10})
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.10 GB -- Worker memory limit: 4.00 GB
```

Finally, if I set `device_memory_limit='3000 MB'` or less, I get:

```
<Client: scheduler='tcp://10.21.85.118:33279' processes=1 cores=1>
distributed.worker - WARNING - Worker's CUDA device is at 93% memory usage. Pausing worker.  Process memory: 1.87 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 1.87 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 93% memory usage. Pausing worker.  Process memory: 1.87 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 1.87 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 206% memory usage. Pausing worker.  Process memory: 4.14 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 4.14 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 147% memory usage. Pausing worker.  Process memory: 2.94 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 2.94 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 39% memory usage. Resuming worker. Process memory: 785.91 MB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 1.43 GB -- Worker memory limit: 2.00 GB
('tcp://10.21.85.118:44729', {'data.device.fast': 0, 'data.device.slow': 2})
distributed.worker - WARNING - Worker's CUDA device is at 84% memory usage. Pausing worker.  Process memory: 1.68 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 1.68 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 93% memory usage. Pausing worker.  Process memory: 1.87 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 1.87 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 109% memory usage. Pausing worker.  Process memory: 2.19 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 2.19 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 110% memory usage. Pausing worker.  Process memory: 2.21 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 2.26 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 122% memory usage. Pausing worker.  Process memory: 2.44 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 2.51 GB -- Worker memory limit: 2.00 GB
('tcp://10.21.85.118:44729', {'data.device.fast': 0, 'data.device.slow': 6})
distributed.worker - WARNING - Worker's CUDA device is at 129% memory usage. Pausing worker.  Process memory: 2.59 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 2.59 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 143% memory usage. Pausing worker.  Process memory: 2.87 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 2.85 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 153% memory usage. Pausing worker.  Process memory: 3.08 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.08 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 154% memory usage. Pausing worker.  Process memory: 3.10 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.10 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - Worker's CUDA device is at 154% memory usage. Pausing worker.  Process memory: 3.10 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.10 GB -- Worker memory limit: 2.00 GB
('tcp://10.21.85.118:44729', {'data.device.fast': 0, 'data.device.slow': 10})
distributed.worker - WARNING - Worker's CUDA device is at 154% memory usage. Pausing worker.  Process memory: 3.10 GB -- Worker memory limit: 2.00 GB
distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 3.10 GB -- Worker memory limit: 2.00 GB
```

We now see a clear distinction between all 3 cases, the first one not spilling the memory at all, the second case spilling memory to host but still running, and the last case never resuming, since there's memory it can't spill. So I believe the core of this PR is correct.

Now, going back to @VibhuJawa's previous comment on this line https://github.com/rapidsai/dask-cuda/pull/35/files#diff-c87f0866b277f959dc7c5d1e4b0ff015R243 I believe he's right. I added some printing of used memory before and after doing `del k, v` and the memory use doesn't decrease, as we would expect. This said, could @mrocklin or @kkraus14 confirm if this the correct way of freeing a chunk of Dask device memory, in particular, related to cudf?

Reply:
> In the end, the client/worker are just idle, with nvidia-smi reporting usage of 2953 MB.
> distributed.worker - WARNING - CUDA device memory use is high but worker has no data to store to host.  Perhaps some other process is leaking memory?  Process memory: 4.14 GB -- Worker memory limit: 5.00 GB

Quick question.  Do we have confidence in our measurement of device memory?

Reply:
> This said, could @mrocklin or @kkraus14 confirm if this the correct way of freeing a chunk of Dask device memory, in particular, related to cudf?

From dask's perspective my guess is that all we can do is call `del v`.  I don't have a strong opinion on how we should be handling memory managers, especially given that there are many of them.

RMM isn't on by default though, so my guess is that @VibhuJawa would know if he was using the RAPIDS memory manager.

Reply:
> Quick question. Do we have confidence in our measurement of device memory?

If you're strictly asking about how much memory is actually allocated on a device (and how much memory is still free at a given time), yes.

> From dask's perspective my guess is that all we can do is call `del v`. I don't have a strong opinion on how we should be handling memory managers, especially given that there are many of them.

My question is more along the lines of, given a Dask chunk (in particular coming from cudf), will calling del on it trigger a CUDA memory deallocation?

> RMM isn't on by default though, so my guess is that @VibhuJawa would know if he was using the RAPIDS memory manager.

I didn't know about that, nevertheless, I tried disabling it, and indeed I didn't see any difference.

Reply:
> My question is more along the lines of, given a Dask chunk (in particular coming from cudf), will calling del on it trigger a CUDA memory deallocation?

The specific answer is, I think, yes.

The general answer is that we probably can't hard-code cudf behavior too much.  This would have to work with whatever GPU enabled library the user decides to use (cupy, pytorch, numba, ...).  

The way for Dask to say it no longer needs something is `del v`.  That's the stanard Python way to tel the GC that you no longer need an object.  If that object allocated something outside of Python then it's up to that library (cudf in this case) to clean things up appropriately.  Dask probably can't get involved here.

Reply:
> The general answer is that we probably can't hard-code cudf behavior too much.

Absolutely agreed. At first I thought there may be something different with the behavior on cudf, but `del v` seems to also not be releasing memory within CuPy. I'm gonna investigate this and see if I can figure out what's going on, hopefully it will be some bug here that has a generic fix no matter what library we're using.

Reply:
I think that it would be interesting to see if this continues to fail in a very simple case where we only use cudf or only use cupy.  

```python
import cupy

memory = []
for i in range(100):
    x = cupy.ones(...)
    memory.append(get_device_memory())
    del x

print(memory)
```

If memory does increase over time, then that's something that we might raise up to the cupy issue tracker.  (my apologies if you're already planning something similar)

Reply:
So, yeah, i think we are creating another shallow copy of the df/cupy somewhere.

I tried the following to delete columns individually (deletes columns in all shallow copies of the data frames ) and the memory is released then. 

```python 

import time
from dask.distributed import Client, wait
from dask_cuda import LocalCUDACluster
import dask_cudf
import cudf
import os
# Use dask-cuda to start one worker per GPU on a single-node system
cluster = LocalCUDACluster(ip='0.0.0.0',n_workers=1)
client = Client(cluster)
# Code to simulate_data


def generate_file(output_file,rows=100):
    with open(output_file, 'wb') as f:
        f.write(b'A,B,C,D,E,F,G,H,I,J,K\n')
        f.write(b'22,697,56,0.0,0.0,0.0,0.0,0.0,0.0,0,0\n'*rows)
        f.close()

# generate the test file 
output_file='test.csv'
generate_file(output_file,rows=50_000_000)


def get_data(dask_worker):
    return {'data.device.fast':len(dask_worker.data.device.fast),
           'data.device.slow':len(dask_worker.data.device.slow),
           }

def remove_data_colwise(dask_worker):
    if dask_worker.data.device.fast:
        k, df, weight = dask_worker.data.device.fast.evict()
        for col in df.columns:
            #pass 
            del df[col] # if you comment this out memory wont be released
        del k
        del df
        return {'data.device.fast':len(dask_worker.data.device.fast),
           'data.device.slow':len(dask_worker.data.device.slow),
               }
    else:
        return {'data:':'No Data'}
    
df = dask_cudf.read_csv('file://'+os.getcwd()+'/'+output_file,chunksize='500 MiB')
df = df.persist()
wait(df)
print(client.run(get_data))
print(client.run(remove_data_colwise))
print(client.run(get_data))
print(client.run(remove_data_colwise))
print(client.run(get_data))
print(client.run(remove_data_colwise))
print(client.run(get_data))
print(client.run(remove_data_colwise))

```
So i will try to look for that.  

Also, yeah ``RMM`` release does not help.

Reply:
Thanks for the suggestion @mrocklin. At first I did something much more "fancy" and didn't realize why it didn't fix the issue. At first I disabled CuPy's memory pool on the client code, which is _not_ the worker, so disabling the memory pool on the client side has no effect (which is now obvious).

Then I tried your suggestion, still didn't work, because I didn't disable the memory pool. Then I disabled the memory pool, and it actually released the memory then.

Here's some code to reproduce what I did:

```python
import cupy
from numba import cuda

cupy.cuda.set_allocator(None)
cupy.cuda.set_pinned_memory_allocator(None)

def get_device_used_memory():
    memory_info = cuda.current_context().get_memory_info()
    return memory_info[1] - memory_info[0]

store = []
used_mem_alloc = []
used_mem_release = []

for i in range(100):
    x = cupy.ones(int(10e6))
    store.append(x)
    used_mem_alloc.append(get_device_used_memory())

while len(store) > 0:
    store.pop()
    used_mem_release.append(get_device_used_memory())

## Alternatively
#for i in range(99, -1, -1):
#    del store[i]
#    used_mem_release.append(get_device_used_memory())


print(used_mem_alloc)
print(used_mem_release[::-1])
assert used_mem_alloc == used_mem_release[::-1]
```

Note that exact values are not necessarily going to match (if there's some other process using the GPU simultaneously), thus the last assert may fail.

Long story short: release CuPy memory via `del` works, but we can't guarantee that unless we disable the memory pool on the worker side. I hacked `CUDAWorker` to disable CuPy's memory pool, and it succeeds to release memory with `del` then. Disabling RMM (which is by default disabled anyway, so probably there was no effect in what I did) failed to release memory.

We have now to provide an interface for users to disable memory pools _on the worker_ for whatever libraries they're using. Alternatively, we could allow them to pass a custom function or something that would call their libraries' garbage collectors. Any ideas/suggestions what would be a better alternative?

Reply:
I think that this is a major issue, but probably not something that we're going to solve in this PR.  The number of memory managers out there is a problem for systems like this that try to manage the GPU for a variety of libraries.  (cc @nouiz)

In the meantime, I recommend that we get this to be good enough and then merge it in.

Reply:
I agree that this is a major issue, that's why I'm concerned with it. In particular, I think pausing is something that can't be enabled under these circumstances, if it is, then when Dask spills memory to host but can't really release that, the worked will get stuck.

I have two proposals (not necessarily mutually exclusive) until we come up with something else:

1. Allow the user to register a function that will be run at the initialization of the worker, to allow the user to disable any memory managers that will be used throughout the runtime;
1. Allow the user to register a function that will be called in `CUDAWorker.device_memory_monitor`, triggering memory managers' release of memory.

We can also disable pausing by default, which I'm inclined to think should be the default to prevent this sort of situation.

Reply:
For 1 see https://github.com/dask/distributed/pull/2453

Disable pausing by default seems fine to me.  This is just a config value change at this point, yes?

Reply:
> For 1 see [dask/distributed#2453](https://github.com/dask/distributed/pull/2453)

Yes, that would be something similar, if not exactly that (sorry, I can't understand all the details without diving in a bit deeper).

Any thoughts on item 2. as well?

> Disable pausing by default seems fine to me. This is just a config value change at this point, yes?

Yes. I just don't know if disabling pause has no other consequences. On the host, I guess this is to prevent the host from running out of memory and eventually getting killed, is that right?

Reply:
> Any thoughts on item 2. as well?

I can see how it would solve the problem.  I guess I'm hoping that medium term it's not necessary.  My inclination is to wait until we have a real-world problem that needs this before adding it.  I won't be surprised if that problem occurs quickly, but I'd still rather put it off and get this in.

Reply:
No objections from me. That said, I have no further changes to be added, from my side, it's ready for more reviews or merging.